My Project/clsTranslateWinForms.vb

```diff
@@ -1 +1,638 @@
+                                                   Optional strParentName As String = "")
+        If IsNothing(clsControl) OrElse IsNothing(clsControl.Controls) OrElse IsNothing(dctComponents) Then
+            Exit Sub
+        End If
+
+        'if control is valid, then add it to the dictionary
+        Dim strControlName As String = ""
+        If Not String.IsNullOrEmpty(clsControl.Name) Then
+            strControlName = If(String.IsNullOrEmpty(strParentName), clsControl.Name, strParentName & "_" & clsControl.Name)
+            If Not dctComponents.ContainsKey(strControlName) Then  'ignore components that are already in the dictionary
+                dctComponents.Add(strControlName, clsControl)
+            End If
+        End If
+
+        For Each ctlChild As Control In clsControl.Controls
+
+            'Recursively process different types of menus and child controls
+            If TypeOf ctlChild Is MenuStrip Then
+                Dim clsMenuStrip As MenuStrip = DirectCast(ctlChild, MenuStrip)
+                GetDctComponentsFromMenuItems(clsMenuStrip.Items, dctComponents)
+            ElseIf TypeOf ctlChild Is ToolStrip Then
+                Dim clsToolStrip As ToolStrip = DirectCast(ctlChild, ToolStrip)
+                GetDctComponentsFromMenuItems(clsToolStrip.Items, dctComponents)
+            ElseIf TypeOf ctlChild Is Control Then
+                GetDctComponentsFromControl(ctlChild, dctComponents, strControlName)
+            End If
+
+        Next
+    End Sub
+
+    '''--------------------------------------------------------------------------------------------
+    ''' <summary>   
+    '''    Populates dictionary <paramref name="dctComponents"/> with all the menu items, and 
+    '''    sub-menu items in the <paramref name="clsMenuItems"/>. 
+    '''    The dictionary can then be used to conveniently translate the menu item text (see other 
+    '''    functions and subs in this class).
+    ''' </summary>
+    '''
+    ''' <param name="clsMenuItems">     The list of menu items to populate the dictionary. </param>
+    ''' <param name="dctComponents">    [in,out] Dictionary to store the menu items. </param>
+    '''--------------------------------------------------------------------------------------------
+    Private Shared Sub GetDctComponentsFromMenuItems(clsMenuItems As ToolStripItemCollection, ByRef dctComponents As Dictionary(Of String, Component))
+        If IsNothing(clsMenuItems) OrElse IsNothing(dctComponents) Then
+            Exit Sub
+        End If
+
+        For Each clsMenuItem As ToolStripItem In clsMenuItems
+
+            'if menu item is valid, then add it to the dictionary
+            If Not (String.IsNullOrEmpty(clsMenuItem.Name) OrElse
+                    dctComponents.ContainsKey(clsMenuItem.Name)) Then 'ignore components that are already in the dictionary
+                dctComponents.Add(clsMenuItem.Name, clsMenuItem)
```

Comment:
Line 308 you mention `'ignore components that are already in the dictionary` yet I'm not seeing,
 `Not dctComponents.ContainsKey(clsMenuItem.Name))`

To me condition 
```
Not (String.IsNullOrEmpty(clsMenuItem.Name) OrElse dctComponents.ContainsKey(clsMenuItem.Name))
```

entails valid menu items that are already contained in the dictionary.  I think you meant the code below,

```
Not (String.IsNullOrEmpty(clsMenuItem.Name) AndAlso Not dctComponents.ContainsKey(clsMenuItem.Name))
```

I may have misunderstood that block.

Reply:
I went through all the possible permutations and I think the original code is correct. However it's confusing so I rewrote as:
```
            'if menu item is valid, then add it to the dictionary
            If String.IsNullOrEmpty(clsMenuItem.Name) OrElse dctComponents.ContainsKey(clsMenuItem.Name) Then
                'ignore components that are already in the dictionary
            Else
                dctComponents.Add(clsMenuItem.Name, clsMenuItem)
            End If
```